### PR TITLE
Support 'OPTIONS' keyword in GPDB external table

### PR DESCRIPTION
**Pull request by kuien**
In case user wanna specify parameters with OPTION other than config
file in LOCATION, it provides flexible and friendly grammar.

Usage:
    CREATE EXTERNAL TABLE xxx LOCATION xxx FORMAT xxx
    [OPTION ( [option_key [AS] 'option_value' [, ...]] )]
    xxx;

As it adds a column 'options' in catalog table 'pg_exttable', we need
to update the catalogversion as well.

NOTE: in this commit we don't implement related parts towards pg_dump
    and alter operations yet.

### CODE REVIEW

File: gpMgmt/bin/gppylib/data/4.3.json [hunk 1]
```diff
@@ -164,6 +164,12 @@
          [ ["enumtypid"], "pg_type", ["oid"] ]
       ]
    },
+   "pg_extension" : {
```
Comment (hlinnaka) [inline gpMgmt/bin/gppylib/data/4.3.json]: This seems unrelated. Yes, we should update the json file, but can you open a separate PR for that? Or just consider this as a review for it, and go ahead and push this as a separate commit.

File: src/backend/commands/exttablecmds.c [hunk 2]
```diff
@@ -761,6 +769,54 @@ transformFormatType(char *formatname)
...
+		char       *val = defGetString(defel);
+
+		/*
+		 * Output "<key> '<val>' ", but replace any space chars in the key
+		 * with meta char (MPP-14467)
+		 */
+		while (*key)
+		{
+		    if (*key == ' ')
+		        appendStringInfoString(&extOptsbuf, "<gpx20>");
```
Comment (hlinnaka) [inline src/backend/commands/exttablecmds.c]: What on earth is this?

File: src/backend/commands/exttablecmds.c [hunk 3]
```diff
@@ -761,6 +769,54 @@ transformFormatType(char *formatname)
...
+		    else
+		        appendStringInfoChar(&extOptsbuf, *key);
+		    key++;
+		}
+
+		appendStringInfo(&extOptsbuf, "='%s' ", val);
+	}
+
+    /* +1 leaves room for sprintf's trailing null */
```
Comment (hlinnaka) [inline src/backend/commands/exttablecmds.c]: Bogus comment

File: src/backend/commands/exttablecmds.c [hunk 4]
```diff
@@ -761,6 +769,54 @@ transformFormatType(char *formatname)
...
+    optionStr = extOptsbuf.data;
+
+    /* convert c string to text datum */
+    result = DirectFunctionCall1(textin, CStringGetDatum(optionStr));
```
Comment (hlinnaka) [inline src/backend/commands/exttablecmds.c]: Suggest using CStringGetTextDatum(optionStr).

File: src/backend/catalog/pg_exttable.c [hunk 5]
```diff
@@ -261,6 +264,14 @@ GetExtTableEntryIfExists(Oid relid)
 	Insist(!isNull);
 	extentry->fmtopts = DatumGetCString(DirectFunctionCall1(textout, fmtopts));
 	
+    /* get the external table options string */
+    options = heap_getattr(tuple,
+                           Anum_pg_exttable_options,
+                           RelationGetDescr(pg_exttable_rel),
+                           &isNull);
+
+    Insist(!isNull);
```
Comment (hlinnaka) [inline src/backend/catalog/pg_exttable.c]: I don't think Insist should be used in new code. Please use Assert(), as in the upstream, or "if (isNull) \n elog(ERROR ...)", if you think there's any chance this might happen in practice, and this isn't performance critical.

File: src/backend/catalog/pg_exttable.c [hunk 6]
```diff
@@ -261,6 +264,14 @@ GetExtTableEntryIfExists(Oid relid)
...
+    extentry->options = DatumGetCString(DirectFunctionCall1(textout, options));
```
Comment (hlinnaka) [inline src/backend/catalog/pg_exttable.c]: Suggest using  TextDatumGetCString()

File: src/backend/commands/exttablecmds.c [hunk 7]
```diff
@@ -291,6 +294,10 @@ DefineExternalRelation(CreateExternalStmt *createExtStmt)
 									   createExtStmt->formatOpts,
 									   list_length(createExtStmt->tableElts),
 									   iswritable);
+    /*
+     * Parse and validate OPTION clause.
+     */
+	optionStr = transformExtOpts(createExtStmt->extOptions);
```
Comment (hlinnaka) [inline src/backend/commands/exttablecmds.c]: check indentation

File: src/backend/commands/exttablecmds.c [hunk 8]
```diff
@@ -761,6 +769,54 @@ transformFormatType(char *formatname)
...
 
+/*
+ * Transform the FORMAT options into a text field. Parse the
+ */
+static Datum
+transformExtOpts(List *extOptions)
+{
+    ListCell   *option;
+    Datum       result;
+    char       *optionStr = NULL;
```
Comment (hlinnaka) [inline src/backend/commands/exttablecmds.c]: Initializing to NULL not needed here. I's better to leave variables uninitialized when it's not needed, because that lets the compiler to catch and warn if the variable is later accidentally used uninitialized, rather than fail at runtime if a NULL is inadvertently passed somewhere. (And in this case, once you take away the initializer, a static analysis tool would tell you that the NULL-check around the pfree() at the end of this function is not needed ;-) )

File: src/backend/commands/exttablecmds.c [hunk 9]
```diff
@@ -761,6 +769,54 @@ transformFormatType(char *formatname)
...
+		while (*key)
+		{
+		    if (*key == ' ')
+		        appendStringInfoString(&extOptsbuf, "<gpx20>");
+		    else
+		        appendStringInfoChar(&extOptsbuf, *key);
+		    key++;
+		}
+
+		appendStringInfo(&extOptsbuf, "='%s' ", val);
```
Comment (hlinnaka) [inline src/backend/commands/exttablecmds.c]: This needs escaping.

File: src/bin/psql/describe.c [hunk 10]
```diff
@@ -1807,7 +1807,7 @@ describeOneTableDetails(const char *schemaname,
 		PGresult   *result;
 
 		printfPQExpBuffer(&buf,
-						  "SELECT x.location, x.fmttype, x.fmtopts, x.command, "
+						  "SELECT x.location, x.fmttype, x.fmtopts, x.options, x.command, "
```
Comment (hlinnaka) [inline src/bin/psql/describe.c]: This needs a version-check guard, so that it still works against a 4.3 server.

File: src/include/catalog/pg_exttable.h [hunk 11]
```diff
@@ -31,6 +31,7 @@ CATALOG(pg_exttable,6040) BKI_WITHOUT_OIDS
 	text	location[1];		/* array of URI strings */
 	char	fmttype;			/* 't' (text) or 'c' (csv) */
 	text	fmtopts;			/* the data format options */
+	text	options;			/* the external table options */
```
Comment (hlinnaka) [inline src/include/catalog/pg_exttable.h]: How about using a text[] for this instead? That's what the upstream uses for FDW options fields.

**Comment by hlinnaka** [comment]
Please also be sure to fix pg_dump before pushing.

File: src/backend/commands/exttablecmds.c (hunk 8 above)
Comment (kuien) [inline src/backend/commands/exttablecmds.c]: Thanks for valuable suggestion.

File: src/backend/commands/exttablecmds.c (hunk 3 above)
Comment (kuien) [inline src/backend/commands/exttablecmds.c]: grrrrr, it should be removed with aped codes, thanks for pointing it out.

File: src/bin/psql/describe.c (hunk 10 above)
Comment (kuien) [inline src/bin/psql/describe.c]: Good catch. Version-check is needed. A quick question: 
is there an elegant way to determine it is GPDB 5? Just like the `if(pset.sversion >= 80300)`, I always think the usage of `isGPDB5000OrLater()` is not sweet.

**Comment by hlinnaka** [comment]
On 12/13/2016 05:59 AM, Kuien Liu wrote:
> kuien commented on this pull request.
>
>
>
>> @@ -1807,7 +1807,7 @@ describeOneTableDetails(const char *schemaname,
>  		PGresult   *result;
>
>  		printfPQExpBuffer(&buf,
> -						  "SELECT x.location, x.fmttype, x.fmtopts, x.command, "
> +						  "SELECT x.location, x.fmttype, x.fmtopts, x.options, x.command, "
>
> Good catch. Version-check is needed. A quick question:
> is there an elegant way to determine it is GPDB 5? Just like the `if(pset.sversion >= 80300)`, I always think the usage of `isGPDB5000OrLater()` is not sweet.

Hmm, good question. This might be the first GPDB-specific new syntax we 
have in 5.0, that needs a version check. All the new stuff from the 8.3 
merge are guarded by 80300, since they're from the upstream.

I think isGPDB5000OrLater() is the best answer for now. It's 
implementation is quite ugly, and it doesn't cache the result, but 
that's not the caller's fault.

- Heikki

**Comment by pf-qiu** [comment]
@hlinnaka hi Heikki, we've fixed all issues you mentioned please review, thank you.

File: src/backend/catalog/pg_exttable.c [hunk 12]
```diff
@@ -261,7 +264,35 @@ GetExtTableEntryIfExists(Oid relid)
...
 	
+    /* get the external table options string */
+    options = heap_getattr(tuple,
+                           Anum_pg_exttable_options,
+                           RelationGetDescr(pg_exttable_rel),
+                           &isNull);
 
+	if (isNull)
+	{
+		Insist(false); /* options list is always populated (url or ON X) */
```
Comment (hlinnaka) [inline src/backend/catalog/pg_exttable.c]: I don't think Insist should be used in new code. It gives a completely unhelpful error message, if you actually hit it.

I'd suggest something like this instead:
```
if (isNull)
  elog(ERROR, "could not find options for external protocol");
```

I see that Insist is used in the similar checks for the other attributes, but nevertheless, let's try to improve things as we go forward. And it seems more likely to hit this check for options than for the other fields, since we're adding the options as a new column now.

File: src/bin/pg_dump/pg_dump.c [hunk 13]
```diff
@@ -9219,7 +9225,22 @@ dumpExternal(TableInfo *tbinfo, PQExpBuffer query, PQExpBuffer q, PQExpBuffer de
...
 
 		/* Now get required information from pg_exttable */
-		if (g_fout->remoteVersion >= 80214)
+		if (gpdb5)
+		{
+			appendPQExpBuffer(query,
+						  "SELECT x.location, x.fmttype, x.fmtopts, x.command, "
+								  "x.rejectlimit, x.rejectlimittype, "
+						      "(SELECT relname "
+						          "FROM pg_class "
```
Comment (hlinnaka) [inline src/bin/pg_dump/pg_dump.c]: Would be good to schema-qualify pg_catalog.pg_class here. It's not strictly required in pg_dump, I guess, but let's be consistent. (The pg_class and pg_exttable references in this same query below are schema-qualified)

File: src/backend/parser/gram.y [hunk 14]
```diff
@@ -4800,6 +4803,34 @@ format_opt_item:
 			}
 			;
 
+ext_option_opt:
+			OPTION ext_option					{ $$ = $2; }
```
Comment (hlinnaka) [inline src/backend/parser/gram.y]: I think it should be "OPTIONS" rather than "OPTION", for consistency with the upcoming FDW and COPY OPTIONS syntax.

File: src/backend/commands/exttablecmds.c [hunk 15]
```diff
@@ -761,6 +772,45 @@ transformFormatType(char *formatname)
...
+		char       *key = defel->defname;
+		char       *val = defGetString(defel);
+		text       *t;
+		Size       len;
+
+                // first 1 for '=', last 1 for '\0'
+		len = VARHDRSZ + strlen(key) + 1 + strlen(val) + 1;
+		t = palloc(len);
+		SET_VARSIZE(t, len);
+		sprintf(VARDATA(t), "%s=%s", key, val);
```
Comment (adam8157) [inline src/backend/commands/exttablecmds.c]: Do we need to escape the string here?

File: src/backend/parser/gram.y (hunk 14 above)
Comment (adam8157) [inline src/backend/parser/gram.y]: +1

File: src/bin/pg_dump/pg_dump.c [hunk 16]
```diff
@@ -9416,6 +9443,48 @@ dumpExternal(TableInfo *tbinfo, PQExpBuffer query, PQExpBuffer q, PQExpBuffer de
 			customfmt = NULL;
 		}
 
+		if (gpdb5)
```
Comment (d) [inline src/bin/pg_dump/pg_dump.c]: I failed to find out *any* test for the upstream code related to `pg_dump`. Can you kindly point at a test case that covers `pg_dump`ing a foreign table? I am specifically interested in seeing the coverage of `pg_options_to_table` and friends

- jesse

**Comment by hlinnaka** [comment]
On 12/28/2016 05:40 AM, Jesse Zhang wrote:
> I failed to find out *any* test for the upstream code related to
> `pg_dump`. Can you kindly point at a test case that covers
> `pg_dump`ing a foreign table? I am specifically interested in seeing
> the coverage of `pg_options_to_table` and friends

pg_dump is not very well covered in the upstream. It is exercised by the 
pg_upgrade test script though (see src/bin/pg_upgrade/test.sh and 
src/bin/pg_upgrade/TESTING). The pg_upgrade test script first runs the 
server regression suite (make installcheck). The tests leave behind 
various objects, on purpose, so that after the regression suite has run, 
the regression database contains a nice selection of tables, functions, 
etc. After that, the pg_upgrade test script sets up a new cluster from 
scratch, and performs a schema-only pg_dump and restore of the cluster.

I'm not sure if the regression database contains any foreign tables, 
though. So I guess my short answer to your question is "I don't know" 
;-). pg_options_to_table() specifically, though, is also used in some of 
the views in information_schema. So e.g. "select * from 
foreign_table_options" would run it.

- Heikki

**Comment by Eulerizeit** [comment]
@hlinnaka since we are now leveraging pg_upgrade should we be looking at adding those regression tests into GPDB?

**Comment by hlinnaka** [comment]
Yeah, that'd be good. The test script essentially performs an upgrade from 5.0 to 5.0. What we have now, can't quite do that, there were some little things that we had not implemented yet, needed for upgrading from 5.0. @danielgustafsson would remember the details. But yes, we should fix those loose ends and start running the pg_upgrade test script as part of installcheck-world.

On 29 December 2016 15:57:18 EET, Robert Eckhardt <notifications@github.com> wrote:
>@hlinnaka since we are now leveraging pg_upgrade should we be looking
>at adding those regression tests into GPDB?
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/greenplum-db/gpdb/pull/1436#issuecomment-269633429

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by danielgustafsson** [comment]
> On 29 Dec 2016, at 15:23, Heikki Linnakangas <notifications@github.com> wrote:
> 
> Yeah, that'd be good. The test script essentially performs an upgrade from 5.0 to 5.0. What we have now, can't quite do that, there were some little things that we had not implemented yet, needed for upgrading from 5.0. @danielgustafsson would remember the details. But yes, we should fix those loose ends and start running the pg_upgrade test script as part of installcheck-world.

I have on my backlog to fix up the last small parts to make 5.0 to 5.0
possible, it’s nothing big and should be fairly quick once I get spare cycles.

cheers ./daniel

**Comment by HaozhouWang** [comment]
Hi @hlinnaka, thanks for your comments. 
we have updated this PR based on your comments, please have a look, thanks.

**Comment by adam8157** [comment]
Going to rebase and merge this since all comments are taken care of, test cases are passed and we verified the function.
